### PR TITLE
VSS: Adding ledger time, todos and integration fix

### PR DESCRIPTION
**Pull request by arc-vmware**
This commit contains the following:
1) Ledger time is added.
2) Added fix for optional last received key in the request of StateSnapshotRequest as this usecase was failing.
3) Added a non blocking and non busy wait version of pop (called popTill) in BasicThreadSafeQueue for better performance.
4) Some configurations were hard coded and were not taken from the file and now made configurable. This was a TODO.

### CODE REVIEW

File: client/concordclient/include/client/concordclient/concord_client.hpp [hunk 1]
```diff
@@ -108,6 +108,13 @@ struct SubscribeConfig {
   std::string pem_private_key;
 };
 
+struct StateSnapshotConfig {
+  // Number of threads available for state snapshot
+  uint32_t num_threads;
+  // Timeout till grpc connection with replicas will wait.
+  uint16_t timeout_in_sec;
```
Comment (dartdart26) [inline client/concordclient/include/client/concordclient/concord_client.hpp:115]: It might make sense to have default values defined here for the members. We can do it in a separate PR.

**Review by dartdart26 — APPROVED** [review]
Looks good, thanks!

File: client/concordclient/include/client/concordclient/concord_client.hpp (hunk 1 above)
Comment (arc-vmware) [reply client/concordclient/include/client/concordclient/concord_client.hpp:115]: thanks @dartdart26 , I will take this in a separate PR.
